modules/top-level/test.nix

```
@@ -50,27 +177,61 @@ in
 
   config =
     let
-      showErr =
-        name: lines:
-        lib.optionalString (lines != [ ]) ''
-          Unexpected ${name}:
-          ${lib.concatStringsSep "\n" (lib.map (v: "- ${v}") lines)}
-        '';
-
-      toCheck =
-        lib.optionalAttrs cfg.checkWarnings { inherit warnings; }
-        // lib.optionalAttrs cfg.checkAssertions { inherit assertions; };
-
-      errors = lib.foldlAttrs (
-        err: name: lines:
-        err + showErr name lines
-      ) "" toCheck;
+      input = {
+        inherit (config) warnings;
+        assertions = builtins.concatMap (x: lib.optional (!x.assertion) x.message) config.assertions;
+      };
+
+      expectationMessages =
+        name:
+        lib.pipe cfg.${name} [
+          (builtins.filter (x: !x.predicate input.${name}))
+          (builtins.map (x: x.message))
+          (builtins.map (msg: if lib.isFunction msg then msg input.${name} else msg))
+          (
+            x:
+            if x == [ ] then
+              null
+            else
+              ''
+                Failed ${toString (builtins.length x)} expectation${lib.optionalString (builtins.length x > 1) "s"}:
+                ${lib.concatMapStringsSep "\n" (line: "- ${line}") x}
+                For ${name}:
+                ${lib.concatMapStringsSep "\n" (line: "- ${line}") input.${name}}
+              ''
+          )
+        ];
+
+      failedExpectations = lib.genAttrs [ "warnings" "assertions" ] expectationMessages;
     in
     {
+      test = {
+        # If checkWarnings or checkAssertions are disabled, ensure the default expectations are overridden
+        assertions = lib.mkIf (!cfg.checkAssertions) [ ];
+        warnings = lib.mkIf (!cfg.checkWarnings) [ ];
+
+        # Predefined expectation predicates available via the `expect` enum-option
+        predefinedExpectationPredicates = {
+          count.predicate = v: l: builtins.length l == v;
+          count.message =
+            v: l: "Expected length to be ${toString v} but found ${toString (builtins.length l)}.";
+          count.valueType = lib.types.ints.unsigned;
```

```suggestion
          count = {
              predicate = v: l: builtins.length l == v;
              message =
                v: l: "Expected length to be ${toString v} but found ${toString (builtins.length l)}.";
              valueType = lib.types.ints.unsigned;
        };
```

I find this a bit easier to read